libvast/src/system/index.cpp

```diff
@@ -1253,66 +1253,81 @@ index(index_actor::stateful_pointer<index_state> self,
+                    [rp](const caf::error& e) mutable {
+                      rp.deliver(e);
+                    });
+              else
+                rp.deliver(result);
             } else {
               // Pick one partition id at random to be "transformed", all the
-              // other ones are "deleted" from the meta index.
-              auto old_partition_id = old_partition_ids.at(0);
-              self
-                ->request(self->state.meta_index, caf::infinite,
-                          atom::replace_v, old_partition_id, new_partition_id,
-                          aps.synopsis)
-                .then(
-                  [self, rp, old_partition_id, new_partition_id,
-                   result](atom::ok) mutable {
-                    self->state.persisted_partitions.insert(new_partition_id);
-                    self
-                      ->request(static_cast<index_actor>(self), caf::infinite,
-                                atom::erase_v, old_partition_id)
-                      .then(
-                        [=](atom::done) mutable {
-                          rp.deliver(result);
-                        },
-                        [=](const caf::error& e) mutable {
-                          rp.deliver(e);
+              // other ones are "deleted" from the meta index. If the new
+              // partition is empty, all partitions are deleted.
+              auto start = size_t{0};
+              if (aps.synopsis) {
+                start = 1;
+                auto old_partition_id = old_partition_ids.at(0);
+                self
+                  ->request(self->state.meta_index, caf::infinite,
+                            atom::replace_v, old_partition_id, new_partition_id,
+                            aps.synopsis)
+                  .then(
+                    [self, rp, old_partition_id, new_partition_id,
+                     result](atom::ok) mutable {
+                      self->state.persisted_partitions.insert(new_partition_id);
+                      self
+                        ->request(static_cast<index_actor>(self), caf::infinite,
+                                  atom::erase_v, old_partition_id)
+                        .then(
+                          [=](atom::done) mutable {
+                            rp.deliver(result);
+                          },
+                          [=](const caf::error& e) mutable {
+                            rp.deliver(e);
+                          });
+                    },
+                    [rp](const caf::error& e) mutable {
+                      rp.deliver(e);
+                    });
+              } else {
+                rp.deliver(result);
+              }
+              for (size_t i = start; i < old_partition_ids.size(); ++i) {
+                auto partition_id = old_partition_ids[i];
+                self
+                  ->request(self->state.meta_index, caf::infinite,
+                            atom::erase_v, partition_id)
```

Comment:
It is not clear to me why we delegate the first erase to the index but the remaining ones to the meta index. Can you add an explanation?

Reply:
Good catch, I think that was an (old) bug, they should all go to the index. @dominiklohmann can you confirm?

Reply:
We went over this, I think this is the correct change to make.

Reply:
We just agreed that this was indeed a bug in a video call.